### DISCUSSION TITLE
Include request's `tools` configuration in OpenAI LLM calls

### DISCUSSION THREAD

**Discussion by ricwo** [langfuse/langfuse#2539] [Ideas]
## Description
For debugging agentic systems, it would be really helpful to include the `tools` field of the LLM call, at least for OpenAI. The LLM _response_ already includes the chosen tool (if any).

## Example
My outgoing OpenAI request looks like this (through LlamaIndex):
```json
{
  "messages": [
    "user: How are you doing?"
  ],
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "query_engine_tool",
        "description": "Useful for running a natural language query\nagainst a knowledge base and get back a natural language response.\n",
        "parameters": {
          "type": "object",
          "properties": {
            "input": {
              "title": "Input",
              "type": "string"
            }
          },
          "required": [
            "input"
          ]
        }
      }
    }
  ],
  "tool_choice": "query_engine_tool"
}
```

Langfuse only captures the following (Tracing -> Generations):
```json
[
    {
        "role": "user",
        "content": "How are you doing?",
        "additional_kwargs": {}
    }
]
```

Langfuse correctly captures the assistant's response including the choice of tool:
```json
{
    "tool_calls": [
        {
            "id": "call_VCTCLBgPo6uSEzv3yWWCnIqR",
            "type": "function",
            "function": {
                "name": "query_engine_tool",
                "arguments": {
                    "input": "How are you doing?"
                }
            }
        }
    ]
}
```

## Request
Might there already be a config parameter that enables this? If not, is there any chance this can be included and captured through `LlamaIndexCallbackHandler`?

**Comment by marcklingen** [comment]
Thanks for sharing, this should be captured, can you open an issue with code to reproduce this and a public link to a trace of this?
